File: homework/Kiet_Tran/lesson_1/range_sum_query_2d.php
<?php

class NumMatrix {
    private $prefixSum = [];
    /**
     * @param Integer[][] $matrix
     */
    function __construct($matrix) {
        $n = count($matrix);
        $m = count($matrix[0]);
        $prefixSum = [];
        for ($i = 1; $i <= $n; $i++) {
            for ($j = 1; $j <= $m; $j++) {
                $prefixSum[$i][$j] = $matrix[$i-1][$j-1] + $prefixSum[$i-1][$j] + $prefixSum[$i][$j-1] - $prefixSum[$i-1][$j-1];
            }
        }
        $this->prefixSum = $prefixSum;
    }
  
    /**
     * @param Integer $row1
     * @param Integer $col1
     * @param Integer $row2
     * @param Integer $col2
     * @return Integer
     */
    function sumRegion($row1, $col1, $row2, $col2) {
        $prefixSum = $this->prefixSum;
        $row1 += 1; $col1 += 1; $row2 += 1; $col2 += 1;
        return $prefixSum[$row2][$col2] - $prefixSum[$row2][$col1-1] - $prefixSum[$row1-1][$col2] + $prefixSum[$row1-1][$col1-1];
    }
}

File: homework/Van_Le/RangeSumQuery2d.cs
// for constructor, let's make a prefix sum as a cache for 2d array
// for each query, use that cache to return the expected result so that the time complexity is O(1)
// time complexity of constructor is O(m.n) for m is the length of the matrix row and n is the length of the matrix column

public class RangeSumQuery2d {
    private int[,] sums;
    
    public NumMatrix(int[][] matrix) 
    {
        if(matrix == null || matrix.Length == 0)
        {
            return;
        }
        
        sums = new int[matrix.Length + 1, matrix[0].Length + 1];
        
        for(int row = 1; row <= matrix.Length; row++)
        {
            for(int column = 1; column <= matrix[row - 1].Length; column++)
            {
                sums[row, column] = sums[row - 1, column] +  sums[row, column - 1]
                     + matrix[row - 1][column - 1] - sums[row - 1, column - 1];
            }
        }
    }
    
    public int SumRegion(int row1, int col1, int row2, int col2) 
    {
        if(sums == null)
        {
            return 0;
        }
        
        return sums[row2 + 1, col2 + 1]  - sums[row1, col2 + 1]  - sums[row2 + 1, col1]  
					+ sums[row1, col1];
    }
}
